File: packages/paygate-python/paygate/analytics/audit_log.py
"""Append-only hash-chained audit log.

Each record includes the SHA-256 of the previous row, giving tamper-evidence
at replay time. Mirrors ``packages/paygate-node/src/analytics/audit-log.ts``.
"""

from __future__ import annotations

import hashlib
import json
import uuid
from dataclasses import asdict, dataclass, field
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

from ..utils.digest import canonical_json
from ..utils.logger import get_logger


@dataclass(frozen=True)
class AuditRecord:
    id: str
    at: str
    actor: str
    action: str
    target: str
    meta: dict[str, Any]
    prev: str
    hash: str


@dataclass
class _PendingRow:
    id: str
    at: str
    actor: str
    action: str
    target: str
    meta: dict[str, Any] = field(default_factory=dict)
    prev: str = "0" * 64


class AuditLogger:
    """Writes ``<YYYY-MM-DD>.ndjson`` files into ``dir``."""

    def __init__(self, *, dir: str | Path, logger: Any | None = None) -> None:
        self._dir = Path(dir)
        self._dir.mkdir(parents=True, exist_ok=True)
        today = datetime.now(timezone.utc).date().isoformat()
        self._file_path = self._dir / f"{today}.ndjson"
        self._prev_hash = "0" * 64
        self._logger = logger or get_logger("paygate.audit")
        self._load_prev_hash()

    def _load_prev_hash(self) -> None:
        if not self._file_path.exists():
            return
        try:
            with self._file_path.open("r", encoding="utf-8") as fh:
                last_hash: str | None = None
                for line in fh:
                    stripped = line.strip()
                    if not stripped:
                        continue
                    try:
                        row = json.loads(stripped)
                    except json.JSONDecodeError:
                        continue
                    if isinstance(row, dict) and isinstance(row.get("hash"), str):
                        last_hash = row["hash"]
                if last_hash is not None:
                    self._prev_hash = last_hash
        except OSError as err:
            self._logger.warn(
                "could not recover audit prev hash; starting fresh",
                err=str(err),
            )

    def append(
        self,
        *,
        actor: str,
        action: str,
        target: str,
        meta: dict[str, Any] | None = None,
    ) -> AuditRecord:
        base = _PendingRow(
            id=str(uuid.uuid4()),
            at=datetime.now(timezone.utc).isoformat(timespec="milliseconds"),
            actor=actor,
            action=action,
            target=target,
            meta=dict(meta) if meta is not None else {},
            prev=self._prev_hash,
        )
        payload = asdict(base)
        hasher = hashlib.sha256()
        hasher.update(self._prev_hash.encode("utf-8"))
        hasher.update(b"|")
        hasher.update(canonical_json(payload).encode("utf-8"))
        row_hash = hasher.hexdigest()
        record = AuditRecord(**payload, hash=row_hash)
        try:
            with self._file_path.open("a", encoding="utf-8") as fh:
                fh.write(json.dumps(asdict(record)) + "\n")
        except OSError as err:
            self._logger.error("audit write failed", err=str(err))
            raise
        self._prev_hash = row_hash
        return record

    @staticmethod
    def verify(file_path: str | Path) -> dict[str, Any]:
        """Re-derive the hash chain from disk. Returns a dict with either
        ``{"ok": True, "rows": int}`` or ``{"ok": False, "brokenAt": int}``.
        """
        path = Path(file_path)
        prev = "0" * 64
        rows = 0
        with path.open("r", encoding="utf-8") as fh:
            for idx, line in enumerate(fh):
                stripped = line.strip()
                if not stripped:
                    continue
                try:
                    row = json.loads(stripped)
                except json.JSONDecodeError:
                    return {"ok": False, "brokenAt": idx}
                base_payload = {
                    "id": row.get("id"),
                    "at": row.get("at"),
                    "actor": row.get("actor"),
                    "action": row.get("action"),
                    "target": row.get("target"),
                    "meta": row.get("meta", {}),
                    "prev": row.get("prev"),
                }
                hasher = hashlib.sha256()
                hasher.update(prev.encode("utf-8"))
                hasher.update(b"|")
                hasher.update(canonical_json(base_payload).encode("utf-8"))
                expected = hasher.hexdigest()
                if expected != row.get("hash") or row.get("prev") != prev:
                    return {"ok": False, "brokenAt": idx}
                prev = row["hash"]
                rows += 1
        return {"ok": True, "rows": rows}


__all__ = ["AuditLogger", "AuditRecord"]

File: packages/limen-node/src/analytics/audit-log.ts
import { createHash, randomUUID } from 'node:crypto';
import { createWriteStream, existsSync, mkdirSync, readFileSync } from 'node:fs';
import path from 'node:path';
import type { Logger } from '../types.js';

/**
 * Append-only hash-chained audit log.  Each record includes the SHA-256 of
 * the previous row, giving tamper-evidence at replay time.
 *
 *    row_hash = SHA-256(prev_hash || canonical_json(row))
 */

export interface AuditRecord {
  readonly id: string;
  readonly at: string;
  readonly actor: string;
  readonly action: string;
  readonly target: string;
  readonly meta: Readonly<Record<string, unknown>>;
  readonly prev: string;
  readonly hash: string;
}

export interface AuditLoggerOptions {
  readonly dir: string;
  readonly logger: Logger;
}

export class AuditLogger {
  private readonly file: string;
  private readonly stream: ReturnType<typeof createWriteStream>;
  private readonly logger: Logger;
  private prevHash = '0'.repeat(64);

  constructor(opts: AuditLoggerOptions) {
    if (!existsSync(opts.dir)) mkdirSync(opts.dir, { recursive: true });
    const name = `${new Date().toISOString().slice(0, 10)}.ndjson`;
    this.file = path.join(opts.dir, name);
    this.stream = createWriteStream(this.file, { flags: 'a' });
    this.logger = opts.logger;
    this.loadPrevHash();
  }

  private loadPrevHash(): void {
    if (!existsSync(this.file)) return;
    try {
      const raw = readFileSync(this.file, 'utf-8');
      const lines = raw.trimEnd().split('\n');
      for (let i = lines.length - 1; i >= 0; i--) {
        const line = lines[i];
        if (!line) continue;
        const parsed = JSON.parse(line) as AuditRecord;
        if (parsed.hash) {
          this.prevHash = parsed.hash;
          break;
        }
      }
    } catch (err) {
      this.logger.warn({ err: (err as Error).message }, 'could not recover audit prev hash; starting fresh');
    }
  }

  append(row: {
    actor: string;
    action: string;
    target: string;
    meta?: Record<string, unknown>;
  }): AuditRecord {
    const base = {
      id: randomUUID(),
      at: new Date().toISOString(),
      actor: row.actor,
      action: row.action,
      target: row.target,
      meta: row.meta ?? {},
      prev: this.prevHash,
    } as const;
    const hash = createHash('sha256')
      .update(this.prevHash)
      .update('|')
      .update(canonical(base))
      .digest('hex');
    const full: AuditRecord = { ...base, hash };
    this.stream.write(`${JSON.stringify(full)}\n`, (err) => {
      if (err) this.logger.error({ err: err.message }, 'audit write failed');
    });
    this.prevHash = hash;
    return full;
  }

  /** Re-derive the hash chain from disk and report any break. */
  static verify(filePath: string): { ok: true; rows: number } | { ok: false; brokenAt: number } {
    const raw = readFileSync(filePath, 'utf-8');
    const lines = raw.trimEnd().split('\n').filter(Boolean);
    let prev = '0'.repeat(64);
    for (let i = 0; i < lines.length; i++) {
      const line = lines[i];
      if (!line) continue;
      const row = JSON.parse(line) as AuditRecord;
      const expected = createHash('sha256')
        .update(prev)
        .update('|')
        .update(
          canonical({
            id: row.id,
            at: row.at,
            actor: row.actor,
            action: row.action,
            target: row.target,
            meta: row.meta,
            prev: row.prev,
          }),
        )
        .digest('hex');
      if (expected !== row.hash || row.prev !== prev) {
        return { ok: false, brokenAt: i };
      }
      prev = row.hash;
    }
    return { ok: true, rows: lines.length };
  }
}

function canonical(v: unknown): string {
  if (v === null || typeof v !== 'object') return JSON.stringify(v);
  if (Array.isArray(v)) return `[${v.map(canonical).join(',')}]`;
  const o = v as Record<string, unknown>;
  const keys = Object.keys(o).sort();
  return `{${keys.map((k) => `${JSON.stringify(k)}:${canonical(o[k])}`).join(',')}}`;
}
